src/ompl/geometric/planners/informedtrees/eitstar/src/RandomGeometricGraph.cpp

```diff
@@ -510,6 +510,11 @@ namespace ompl
                 {
                     do  // Sample randomly until a valid state is found.
                     {
+                        if (terminationCondition)
+                        {
+                            // We've been asked to stop.
+                            return nullptr;
+                        }
```

Comment:
We have to be careful here to sample the state before checking the termination condition. If the user passes a condition that always terminates (e.g., to run EIT* iteration by iteration), then this change will prevent EIT* from making any progress. I think we could do something like this:

```cpp
std::shared_ptr<State>
RandomGeometricGraph::getNewSample(const ompl::base::PlannerTerminationCondition &terminationCondition)
{
    // Allocate a new state.
    auto state = std::make_shared<State>(spaceInfo_, objective_);

    // Create the requested number of new states.
    if (currentNumSamples_ < buffer_.size())
    {
        // This does at the moment not deal with the fact that samples might become invalid
        // also, these samples might not be informed samples
        state = buffer_[currentNumSamples_];
    }
    else
    {
        bool foundValidSample = false;
        do  // Sample randomly until a valid state is found.
        {
            if (isMultiqueryEnabled_)
            {
                // If we are doing multiquery planning, we sample uniformly, and reject samples that can
                // not improve the solution. This means that we need to sample the whole space, and add
                // the samples to the buffer
                sampler_->sampleUniform(state->raw(), objective_->infiniteCost());
            }
            else
            {
                // In case we are not doing multiquery planning, we can still directly sample the informed
                // set.
                sampler_->sampleUniform(state->raw(), solutionCost_);
            }

            // Keep track of how many states we sampled.
            ++numSampledStates_;

            // Check if the sample is valid.
            foundValidSample = spaceInfo_->isValid(state->raw());
        } while (!foundValidSample && !terminationCondition);

        // The sample is invalid. We have to return to respect the termination condition.
        if (!foundValidSample)
        {
            return nullptr;
        }

        // We've found a valid sample, let's add it to the buffer.
        ++numValidSamples_;
        buffer_.emplace_back(state);
    }
    ++currentNumSamples_;
    return state;
}
```
Not the most elegant code, but hopefully it fixes the issue you see. Would you mind testing this version with your setup?

Reply:
@marlinstrub 
I added the requested changes in [da49a1c](https://github.com/ompl/ompl/pull/994/commits/da49a1c6550100db6302b01be8359e9be15c263b)

To check if it is working as intended, I created a almost impossible to sample environment and ran the code with and without this PR fixes.
![image](https://user-images.githubusercontent.com/4281771/219997122-3af532cf-bcf0-4276-9909-82798ee284f1.png)

### AS IS
- max_planning_time: 5s
```shell
[ WARN] [1676860730.860321117]: ParallelPlan::solve(): Unable to find solution by any of the threads in 7.615564 seconds
[ WARN] [1676860730.860351988]: Timed out
[ WARN] [1676860730.866058622]: Fail: ABORTED: TIMED_OUT
[ WARN] [1676860743.322497691]: ParallelPlan::solve(): Unable to find solution by any of the threads in 8.275939 seconds
[ WARN] [1676860743.322526399]: Timed out
[ WARN] [1676860743.338020238]: Fail: ABORTED: TIMED_OUT
[ WARN] [1676860752.686450662]: ParallelPlan::solve(): Unable to find solution by any of the threads in 7.840804 seconds
[ WARN] [1676860752.686479757]: Timed out
```

### TO BE
- max_planning_time: 5s
```shell
[ WARN] [1676861034.370784561]: ParallelPlan::solve(): Unable to find solution by any of the threads in 5.000077 seconds
[ WARN] [1676861034.370829586]: Timed out
[ WARN] [1676861034.382284010]: Fail: ABORTED: TIMED_OUT
[ WARN] [1676861042.756183926]: ParallelPlan::solve(): Unable to find solution by any of the threads in 5.000095 seconds
[ WARN] [1676861042.756211490]: Timed out
[ WARN] [1676861042.773767959]: Fail: ABORTED: TIMED_OUT
[ WARN] [1676861049.156938698]: ParallelPlan::solve(): Unable to find solution by any of the threads in 5.000158 seconds
[ WARN] [1676861049.156962079]: Timed out
[ WARN] [1676861049.157320775]: Fail: ABORTED: TIMED_OUT
```